src/threading/MsgThread.cc

```diff
@@ -179,18 +245,20 @@ MsgThread::MsgThread() : BasicThread(), queue_in(this, nullptr), queue_out(nullp
     child_finished = false;
     child_sent_finish = false;
     failed = false;
-    thread_mgr->AddMsgThread(this);
-
-    if ( ! iosource_mgr->RegisterFd(flare.FD(), this) )
-        reporter->FatalError("Failed to register MsgThread fd with iosource_mgr");
 
-    SetClosed(false);
+    io_source = new detail::MsgThread_IOSource(this);
+    thread_mgr->AddMsgThread(this, io_source);
```

Comment:
If `MsgThread_IOSource` becomes a `unique_ptr` inside the manager (see other comment), either use `make_unique()` plus `std::move()` here, or move the instantiation to the manager, too, by replacing `AddMsgThread` with, say, `MsgThread_IOSource* CreateMsgThread(this)` and putting the `make_unique` inside `CreateMsgThread` and they just have it return a non-owning raw pointer.